pkg/util/winutil/crashreport/crashreport.go

```diff
@@ -0,0 +1,144 @@
+
+//go:build windows
+
+// Package crashreport provides shared helpers for recording crash detection state
+package crashreport
+
+import (
+	"fmt"
+
+	sysconfig "github.com/DataDog/datadog-agent/cmd/system-probe/config"
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/system/wincrashdetect/probe"
+	dd_config "github.com/DataDog/datadog-agent/pkg/config"
+	process_net "github.com/DataDog/datadog-agent/pkg/process/net"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/retry"
+
+	"golang.org/x/sys/windows/registry"
+)
+
+// WinCrashReporter is the helper object for getting/storing crash state
+type WinCrashReporter struct {
+	hive             registry.Key
+	baseKey          string
+	startupWarnCount int
+	hasRunOnce       bool
+}
+
+const (
+	maxStartupWarnings = 20
+	reportedKey        = `lastReported`
+)
+
+// NewWinCrashReporter creates the object for checking/setting the windows
+// crash registry keys
+func NewWinCrashReporter(hive registry.Key, key string) (*WinCrashReporter, error) {
+	wcr := &WinCrashReporter{
+		hive:    hive,
+		baseKey: key,
+	}
+	return wcr, nil
+}
+
+func (wcr *WinCrashReporter) generateReportedValue(wcs probe.WinCrashStatus) string {
+	return fmt.Sprintf("%s_%s", wcs.FileName, wcs.DateString)
+}
+
+func (wcr *WinCrashReporter) haveAlreadyReported(wcs probe.WinCrashStatus) bool {
+
+	newval := wcr.generateReportedValue(wcs)
+
+	k, err := registry.OpenKey(wcr.hive, wcr.baseKey, registry.QUERY_VALUE)
+	if err != nil {
+		// key not even there
+		return false
+	}
+	defer k.Close()
+	reportedval, _, err := k.GetStringValue(reportedKey)
+	if err != nil {
+		return false
+	}
+	if newval == reportedval {
+		return true
+	}
+	return false
+}
+
+func (wcr *WinCrashReporter) setReported(wcs probe.WinCrashStatus) {
+	newval := wcr.generateReportedValue(wcs)
+
+	k, _, err := registry.CreateKey(wcr.hive, wcr.baseKey, registry.ALL_ACCESS)
+	if err != nil {
+		// key not even there
+		return
+	}
+	defer k.Close()
+	// if we can't set the value, there's nothing we can do.  On next agent
+	// start, the same crash will be reported if the file is still there.
+	_ = k.SetStringValue(reportedKey, newval)
+}
+
+func (wcr *WinCrashReporter) handleStartupError(err error) error {
+	if retry.IsErrWillRetry(err) {
+		wcr.startupWarnCount++
+		// this is an expected error, the check run will occur before the system probe
+		// comes up.  However, it should resolve.  If the number of these exceeds a
+		// given threshold, report as an error anyway
+		if wcr.startupWarnCount > maxStartupWarnings {
+			return err
+		}
+		return nil
+	}
+	return err
+}
+
+// CheckForCrash uses the system probe crash module to check for a crash
+func (wcr *WinCrashReporter) CheckForCrash() (*probe.WinCrashStatus, error) {
+	if wcr.hasRunOnce {
+		return nil, nil
+	}
+	sysProbeUtil, err := process_net.GetRemoteSystemProbeUtil(
+		dd_config.SystemProbe.GetString("system_probe_config.sysprobe_socket"))
```

Comment:
ditto other `config.SystemProbe` comment. Can this be passed in via the constructor to `WinCrashReporter` instead of using a global?

Reply:
WinCrashReporter is currently being used by a non-component check.